cuqi/geometry/_geometry.py

```diff
@@ -702,32 +702,47 @@ def _compute_eigpairs(self, xnod, C_nu, sigma, M, N_GL):
 
 class StepExpansion(Continuous1D):
     '''
-    class representation of the step random field with 3 intermidiate steps
+    Class representation of the step random field with n_steps intermediate steps.
+
+    Parameters:
+    -----------
+    grid: ndarray
+        Grid points for the step expansion to be evaluated at. There can be more grid points than number of steps.
+
+    n_steps: int
+        Number of equidistant steps.
+
+    kwargs: keyword arguments
+        keyword arguments are passed to the initializer of :class:`~cuqi.geometry.Continuous1D`
     '''
-    def __init__(self, grid, axis_labels=['x'], **kwargs):
+    def __init__(self, grid, n_steps=3, **kwargs):
 
-        super().__init__(grid, axis_labels,**kwargs)
+        super().__init__(grid, **kwargs)
+        if n_steps > np.size(self.grid):
+            raise ValueError("n_steps must be smaller than the number of grid points")
 
+        self._n_steps = n_steps
         L = self.grid[-1]
-        self._idx1 = np.where( (self.grid>=0*L)&(self.grid<=0.333*L) )
-        self._idx2 = np.where( (self.grid>0.333*L)&(self.grid<=0.666*L) )
-        self._idx3 = np.where( (self.grid>0.666*L)&(self.grid<=L) )
+
+        self._indices = []
+        for i in range(self._n_steps):
+            start = i*L/self._n_steps
+            end = (i+1)*L/self._n_steps
+            self._indices.append( np.where((self.grid>=start)&(self.grid<=end))[0] )
 
     def par2fun(self, p):
         real = np.zeros_like(self.grid)  
- 
-        real[self._idx1[0]] = p[0]
-        real[self._idx2[0]] = p[1]
-        real[self._idx3[0]] = p[2]
+        for i in range(self._n_steps):
+            real[self._indices[i]] = p[i]
  
         return real
 
-    def fun2par(self,f):       
-        val1 = f[self._idx1[0][0]]
-        val2 = f[self._idx2[0][0]]
-        val3 = f[self._idx3[0][0]]
-        return np.array([val1,val2,val3])
+    def fun2par(self,f):
+        val = np.zeros(self._n_steps)
```

Comment:
I think I may not be able to review the functionality until the documentation has been updated to clarify what grid points, steps etc are. I would like to return to the review then.

Reply:
Sounds great. I updated the documentation and it will be great if you can take a look and let me know what you think.